src/pb_stub.cc

```diff
@@ -712,13 +715,15 @@ main(int argc, char** argv)
 
   stub->Initialize(model_version, argv[6] /* triton install path */);
   bool non_graceful_exit = false;
+  bool &health = stub->Health();
 
   bool background_thread_running = true;
   std::thread background_thread(
-      [&parent_pid, &background_thread_running, &stub, &non_graceful_exit] {
+      [&parent_pid, &background_thread_running, &stub, &non_graceful_exit, &health] {
         while (background_thread_running) {
-          // Every two seconds check if the parent process is alive.
+          // Every 10ms check if the parent process is alive.
           sleep(0.01);
+          health = true;
```

Comment:
You are communicating between threads using "plain" variables without mutexes or atomics. This one is in shared memory and 'background_thread_running' and others are on stack. There is no guarantee if/when other threads will see this changes. On x86 it might just work but that may not be true on ARM. You need to make there variables atomics or protect them with a mutex or some other sync that enforces memory ordering/visibility.
https://stackoverflow.com/questions/3114124/in-c-c-are-volatile-variables-guaranteed-to-have-eventually-consistent-semant
https://stackoverflow.com/questions/34382440/shared-variables-in-c11

Reply:
I see, I will fix this. I will add a mutex only in the shared memory for the health variable and make the variables that are shared between the threads atomic.